### ISSUE TITLE
Bazel should add Content-MD5 header to remote PUT requests

### ISSUE THREAD

**Issue by allada** [bazelbuild/bazel#10070]
### Description of the problem / feature request:

We had an incident recently where our remote AC cache was tainted by (probably) [an extremely rare TCP/IP checksum false positive incident](http://ccr.sigcomm.org/archive/1995/conf/partridge.pdf). We already do check for data size matches, but Action Cache records (from what I can tell) are not checksumable. This could have been avoided if a `Content-MD5` header was sent by bazel with the PUT request.

I'd be happy to make a patch, however from what I remember accepting patches from external contributors is not trivial due to Google3 being the upstream (unless things have changed in the last few years).

In any case here's a spot in the code that I believe would be a good place to start:
https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/remote/http/HttpUploadHandler.java#L100

### What's the output of `bazel info release`?

We use 0.26.1, but I snooped around on v1.0 source and still could not find the header tag being created anywhere.

**Comment by buchgr** [comment]
Thanks for reporting the issue. That sounds scary. I'd be interested in how you came to your conclusion?

**Comment by buchgr** [comment]
Do you by any chance happen to use bazel-remote? There recently was a bug fix for an upload corruption: https://github.com/buchgr/bazel-remote/commit/900d9dff49a5790073e59bc0a127b44bf7513fb4

**Comment by allada** [comment]
@buchgr, yes I was the one who submitted that PR and this is how I found the bug, but when we checked the size of the AC content that it expected it was correct, but the end of the content was garbage (I guess I could have tried to CRC32 the file manually, but priorities).

In any case, this was the best solution we could come up with. When we ran the numbers on what the chances of this are, we found that it was on the order of 0.1% per year for the volume of data we use. Especially since this happened on a 10 gig local network (ie: no secondary hash or encryption).
